### PR TITLE
[Gemini] Prefetch next chunk before each op

### PR DESCRIPTION
**Pull request by botbw**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by botbw** [comment]
# Profiling
## Prefetch = 0
image

## Prefetch = 10
image

# Benchmart

To add

**Comment by Hz188** [comment]
# LLaMa Benchmark

image

File: colossalai/zero/gemini/placement_policy.py [hunk 1]
```diff
@@ -95,6 +107,24 @@ def setup_grads_device(
...
+            return []
+        can_prefetch = self.max_prefetch - len(async_works)
+        prefetch = []
+        for i in range(compute_idx + 1, len(compute_list)):
+            for chunk in compute_list[i]:
+                if len(prefetch) >= can_prefetch:
+                    break
+                if chunk not in prefetch and chunk not in self.chunk_manager.accessed_chunks:
+                    prefetch.append(chunk)
+            else:
```
Comment (ver217) [inline colossalai/zero/gemini/placement_policy.py:123]: Where its pair `if`?

File: examples/language/gpt/gemini/commons/utils.py [hunk 2]
```diff
@@ -30,8 +30,9 @@ def get_profile_context(enable_flag, warmup_steps, active_steps, save_dir):
             activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
             schedule=schedule(wait=0, warmup=warmup_steps, active=active_steps),
             on_trace_ready=tensorboard_trace_handler(save_dir),
-            record_shapes=True,
-            profile_memory=True,
+            # record_shapes=True,
+            # profile_memory=True,
```
Comment (ver217) [inline examples/language/gpt/gemini/commons/utils.py:34]: why not record shapes

Comment (ver217) [reply examples/language/gpt/gemini/commons/utils.py:34]: It seems it's dupilcated as examples/language/performance_evaluator.py

File: colossalai/zero/gemini/placement_policy.py (hunk 1 above)
Comment (Hz188) [reply colossalai/zero/gemini/placement_policy.py:123]: `else` match with `for` loop, to break out of the outer for loop

File: examples/language/gpt/gemini/commons/utils.py (hunk 2 above)
Comment (botbw) [reply examples/language/gpt/gemini/commons/utils.py]: > It seems it's dupilcated as examples/language/performance_evaluator.py

Resolved

This was previously only used for gpt and is now moved to `performance_evaluator.py` for llama.

**Review by ver217 — APPROVED** [review]

**Pull request merged**